lambda/trmc.ml

```diff
@@ -0,0 +1,515 @@
+          let ls = List.map (traverse ctx) ls in
+          Return (Lstaticraise (id, ls))
+      | Ltrywith (l1, id, l2) ->
+          (* in [try l1 with id -> l2], the term [l1] is
+             not in tail-call position (after it returns
+             we need to remove the exception handler),
+             so it is not transformed here *)
+          let l1 = traverse ctx l1 in
+          lift ctx [l2]
+            (fun [l2] -> Ltrywith (l1, id, l2))
+      | Lstaticcatch (l1, ids, l2) ->
+          (* In [static-catch l1 with ids -> l2],
+             the term [l1] is in fact in tail-position *)
+          lift ctx [l1; l2]
+            (fun [l1; l2] -> Lstaticcatch (l1, ids, l2))
+      | Levent (lam, lev) ->
+          lift ctx [lam]
+            (fun [lam] -> Levent (lam, lev))
+      | Lifused (x, lam) ->
+          lift ctx [lam]
+            (fun [lam] -> Lifused (x, lam))
+    end
+
+  (* [lift ctx tail_terms context] optimizes a term of the form
+     C[t1,..,tn] where the t1,..,tn are subterms of the multi-context C
+     that are all in tail position.
+
+     It works by recursively compiling each t1..tn into the corresponding choice.
+     If they are all Return, we Return the overall context;
+     otherwise there is at least one tail-term
+     that is Set (would benefit from TRMC), so we Set.
+  *)
+  and lift ctx tail_terms context =
+    let choices = List.map (choice ctx) tail_terms in
+    match find_settable choices with
+    | Error all_returns ->
+        Return (context all_returns)
+    | Ok _ ->
+        let noloc = Debuginfo.Scoped_location.Loc_unknown in
+        Set {
+          dps = (fun dst -> context (List.map (set noloc dst) choices));
+          direct = (fun () -> context (List.map return choices));
+        }
+
+  and choice_apply ctx apply =
+    let exception No_trmc in
+    try
+      match apply.ap_func with
+      | Lvar f ->
+          (* TODO: if [@tailcall false] then raise No_trmc; *)
+          let candidate =
+            try Ident.Map.find f ctx.candidates
+            with Not_found ->
+              (* TODO warn: tail-callness of the call is broken in
+                 the destination-passing-style version; either the function [f]
+                 should be marked as trmc-specializable at the callsite,
+                 or the user should add [@tailcall false] to clarify
+                 that they are aware of this limitation. *)
+              raise No_trmc
+          in
+          Set {
+            dps = (fun dst ->
+              let f_dps = candidate.dps_id in
+              Lapply { apply with
+                       ap_func = Lvar f_dps;
+                       ap_args = add_dst_args dst apply.ap_args;
+                       ap_tailcall = Should_be_tailcall;
+                     });
+            direct = (fun () -> Lapply apply);
+          }
+      | _nontail -> raise No_trmc
+    with No_trmc -> Return (Lapply apply)
+
+  and choice_makeblock ctx (tag, flag, shape) blockargs loc =
```

Comment:
Still need to wrap my head around how this works exactly, but it looks correct.  I will re-read it after lunch.